src/sentry/spans/buffer_v2.py

```diff
@@ -0,0 +1,290 @@
+* The span buffer is sharded by project. Therefore, condition 4 is handled by the code for condition 3, although with some delay.
+
+Segments are flushed out to `buffered-spans` topic under two conditions:
+
+* If the segment has a root span, it is flushed out after `span_buffer_root_timeout` seconds of inactivity.
+* Otherwise, it is flushed out after `span_buffer_timeout` seconds of inactivity.
+
+Now how does that look like in Redis? For each incoming span, we:
+
+1. Try to figure out what the name of the respective span buffer is (`set_key` in `add-buffer.lua`)
+  a. We look up any "redirects" from the span buffer's parent_span_id (key "span-buf:sr:{project_id:trace_id}:span_id") to another key.
+  b. Otherwise we use "span-buf:s:{project_id:trace_id}:span_id"
+2. Rename any span buffers keyed under the span's own span ID to `set_key`, merging their contents.
+3. Add the ingested span's payload to the set under `set_key`.
+4. To a "global queue", we write the set's key, sorted by timeout.
+
+Eventually, flushing cronjob looks at that global queue, and removes all timed
+out keys from it. Then fetches the sets associated with those keys, and deletes
+the sets.
+
+This happens in two steps: Get the to-be-flushed segments in `flush_segments`,
+then the consumer produces them, then they are deleted from Redis
+(`done_flush_segments`)
+
+On top of this, the global queue is sharded by partition, meaning that each
+consumer reads and writes to shards that correspond to its own assigned
+partitions. This means that extra care needs to be taken when recreating topics
+or using spillover topics, especially when their new partition count is lower
+than the original topic.
+"""
+
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import NamedTuple
+
+import rapidjson
+from django.conf import settings
+from sentry_redis_tools.clients import RedisCluster, StrictRedis
+
+from sentry.utils import metrics, redis
+
+# This SegmentId is an internal identifier used by the redis buffer that is
+# also directly used as raw redis key. the format is
+# "span-buf:s:{project_id:trace_id}:span_id", and the type is bytes because our redis
+# client is bytes.
+#
+# The segment ID in the Kafka protocol is actually only the span ID.
+SegmentId = bytes
+
+
+def get_redis_client() -> RedisCluster[bytes] | StrictRedis[bytes]:
+    return redis.redis_clusters.get_binary(settings.SENTRY_SPAN_BUFFER_CLUSTER)
+
+
+add_buffer_script = redis.load_redis_script("spans/add-buffer.lua")
+
+
+def segment_to_span_id(segment_id: SegmentId) -> bytes:
+    return segment_id.rsplit(b":", 1)[1]
+
+
+# fun fact: namedtuples are faster to construct than dataclasses
+class Span(NamedTuple):
+    trace_id: str
+    span_id: str
+    parent_span_id: str | None
+    project_id: int
+    payload: bytes
+    is_segment_span: bool = False
+
+
+class RedisSpansBufferV2:
```

Comment:
Could we remove the "V2" from the class name and file name? It doesn't provide any additional information, there's no V1 in code anymore, and it's the first version to go live. 

Also, the old file structure made more sense to me. The class name suggests there can be more implementations conforming to the same interface. Either rename the class to `SpanBuffer` (or `SpansBuffer`), or move this back into `buffer/redis.py` to show clear intent in the design.

Reply:
I renamed the class